packages/core/src/shared/sam/utils.ts

```diff
@@ -108,6 +109,38 @@ export async function updateRecentResponse(
         getLogger().warn(`sam: unable to save response at key "${key}": %s`, err)
     }
 }
+
+/**
+ * Returns true if there's an ongoing build process for the provided template, false otherwise
+ * @Param templatePath The path to the template.yaml file
+ */
+function isBuildInProgress(templatePath: string): boolean {
+    const expirationDate = getRecentResponse(buildProcessMementoRootKey, globalIdentifier, templatePath)
+    if (expirationDate) {
+        return Date.now() < parseInt(expirationDate)
+    }
+    return false
+}
+
+/**
+ * Throws an error if there's a build in progress for the provided template
+ * @Param templatePath The path to the template.yaml file
+ */
+export function throwIfTemplateIsBeingBuilt(templatePath: string) {
+    if (isBuildInProgress(templatePath)) {
+        throw new ToolkitError('This template is already being built', { code: 'BuildInProgress' })
+    }
+}
+
+export async function registerTemplateBuild(templatePath: string) {
```

Comment:
"registering" is usually a sign of a design flaw. registries require lifecycle management, and can get "corrupted" by logic bugs (like any "cache" in general).

Is there instead a way we can check SAM CLI's state (e.g. in the `.sam-...` directory) to see if it's currently building something?

Reply:
We currently can't check sam cli state in the toolkit. I think this is the best we can do right now, until we find a solution in SAM CLI.